### PR TITLE
USe asyncmap and avoid request copy

### PR DESCRIPTION
**Pull request by lonelycode**
This moves the `orgChanMap` into a`sync.Map` and removes reads from the request object by the AccessNext function, but retains the original request pointer in order to handle the context write required for the proxy at the end of the chain.

By having the request copy operation in there, the request is being copied for every request coming into the gateway. which potentially could impact throughput.

### CODE REVIEW

**Comment by lonelycode** [comment]
Note this also removes the request encoding in the event handler for simplicity sake, I feel it's a fair trade-off for async.

File: mw_organisation_activity.go [hunk 1]
```diff
@@ -134,7 +126,7 @@ func (k *OrganizationMonitor) ProcessRequestOffThread(r *http.Request) (error, i
 	return nil, 200
 }
 
-func (k *OrganizationMonitor) AllowAccessNext(orgChan chan bool, r *http.Request) {
+func (k *OrganizationMonitor) AllowAccessNext(orgChan chan bool, path string, IP string, r *http.Request) {
```
Comment (letzya) [inline mw_organisation_activity.go]: If IP and path are independent of the request then it makes sense to add them the function signature but if they are always part of it then why not then why did you add them (they can be set locally in the function, unless I misunderstood the purpose)

File: log_helpers.go [hunk 2]
```diff
@@ -30,3 +30,23 @@ func getLogEntryForRequest(r *http.Request, key string, data map[string]interfac
...
+
+func getExplicitLogEntryForRequest(path string, IP string, key string, data map[string]interface{}) *logrus.Entry {
+	// populate http request fields
+	fields := logrus.Fields{
+		"path":   path,
+		"origin": IP,
+	}
+	// add key to log if configured to do so
+	if key != "" {
+		fields["key"] = key
```
[the remark this answers is not in the document]
Comment (lonelycode) [reply log_helpers.go]: This isn't my code - i just duplicated the previous function (see the same file above), all I did was remove the request object requirements, can't speak to the actual contents

File: mw_organisation_activity.go (hunk 1 above)
Comment (lonelycode) [reply mw_organisation_activity.go]: Because the request object was causing a read panic later in the stack - see the convo in slack

File: log_helpers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (letzya) [reply log_helpers.go]: 👍

[the remark this answers is not in the document]
Comment (letzya) [reply log_helpers.go]: 👍
